rust/src/storage/s3/dynamodb.rs

```diff
@@ -0,0 +1,427 @@
+use std::collections::HashMap;
+use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};
+
+use rusoto_core::RusotoError;
+use rusoto_dynamodb::*;
+use uuid::Uuid;
+
+pub struct DynamoLock {
+    client: DynamoDbClient,
+    opts: Options,
+}
+
+#[derive(Clone, Debug)]
+pub struct Options {
+    pub partition_key_value: String,
+    pub table_name: String,
+    pub owner_name: String,
+    pub lease_duration: u64,
+    pub sleep_millis: u64,
+}
+
+#[derive(Clone, Debug)]
+pub struct LockItem {
+    pub owner_name: String,
+    pub record_version_number: String,
+    pub lease_duration: u64,
+    pub is_released: bool,
+    pub data: Option<String>,
+    pub lookup_time: u128,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum DynamoError {
+    #[error("Dynamo table not found")]
+    TableNotFound,
+
+    #[error("Conditional check failed")]
+    ConditionalCheckFailed,
+
+    #[error("DynamoDB item has invalid schema")]
+    InvalidItemSchema,
+
+    #[error("Could not acquire lock for {0} sec")]
+    TimedOut(u64),
+
+    #[error("Put item error: {0}")]
+    PutItemError(RusotoError<PutItemError>),
+
+    #[error("Update item error: {0}")]
+    UpdateItemError(#[from] RusotoError<UpdateItemError>),
+
+    #[error("Get item error: {0}")]
+    GetItemError(RusotoError<GetItemError>),
+
+    #[error("Delete item error: {0}")]
+    DeleteItemError(#[from] RusotoError<DeleteItemError>),
+}
+
+pub const PARTITION_KEY_NAME: &str = "key";
+pub const OWNER_NAME: &str = "ownerName";
+pub const RECORD_VERSION_NUMBER: &str = "recordVersionNumber";
+pub const IS_RELEASED: &str = "isReleased";
+pub const LEASE_DURATION: &str = "leaseDuration";
```

Comment:
have you looked into using item ttl to manage lock expiration instead? i am assuming we are manually managing the expiration here using lease_duration because that's what the reference java implementation did?

Reply:
the benefit of ttl is it could simplify the design, where expired locks will get deleted in dynamodb automatically so we can delete all the expiration checking logic in our code. if a lock item exists in dyanmodb, it's a valid and active lock, otherwise we are safe to create a new lock.

Reply:
ttl is not reliable since it could be up to one day between dynamodb triggers deletion and item ttl's passes. That's what originally is used in rust's dynalock and as it turned out was not right way for our case.  Also we use update_item to set is_released=true (instead of deleting) for the lock releasing which allows us to store some arbitrary metadata within lock item such us latest delta log version or kafka offsets for kafka-delta-ingest

Reply:
ha, for some reason i had the impression that dynamodb honors TTL in a synchronous and consistent fashion, that's unfortunate :(